### ISSUE TITLE
[HybridApp] Clicking expense violation notifications cause the iOS app to crash

### ISSUE THREAD

**Issue by Julesssss** [Expensify/App#50423]
@AndrewGable @trjExpensify 

### Problem

Clicking expense violation notifications cause the iOS HybridApp to crash. NewDot users should not be receiving OldDot notifications.

**Reproduction**
_I only have HybridApp installed, I think it’s:_
- Have a workspace on New Expensify that has violations enabled
- SmartScan a receipt that has violations (e.g. our limit is $10 and this SmartScan was $100)
- Click on push

![IMG_7739](https://github.com/user-attachments/assets/0f527d28-e9be-40ce-b32c-59b11248ec8a)

### Solution

Figure out if this should block the upcoming HybridApp release.

Fix.

**Comment by Julesssss** [comment]
I'm struggling to locate the crash in Firebase. 

The higher frequency crashes exist in pre-hybrid releases, or seem very unrelated ([shadowview example](https://console.firebase.google.com/project/expensify-mobile-app/crashlytics/app/ios:com.expensify.expensifylite/issues/754e3902437176e57ff794418fe4f3fb?time=last-seven-days&types=crash&versions=9.0.39%20(9.0.39.0);9.0.39%20(9.0.39.5)&sessionEventKey=90e177efce9d4b2e952acf85d5dec450_2001857906433873323)).

**Comment by AndrewGable** [comment]
I have been reporting this via the TestFlight crash dialog, I am not sure Firebase is initialized yet

**Comment by Julesssss** [comment]
So for now would our desired solution to be stop sending the notification for NewDot violations, rather than implementing NewDot violation handling? 

Secondly, is the desired behaviour to only send push notifications for the experience the user is set to? For example, a NewDot user would not receive a notification for an OldDot violation, and a NewDot user would not (currently) receive any violation notification.

**Comment by AndrewGable** [comment]
I think we want to stop sending and we will figure out how to handle violation push notifications in a different solution. I think we are discussing them here: https://expensify.slack.com/archives/C06ML6X0W9L/p1728408150428299

**Comment by trjExpensify** [comment]
What's a "NewDot violation", and what's an "OldDot violation", just so I'm clear?

I'm not aware of a system message to notify when a receipt upload contains workspace violations in NewDot. @JmillsExpensify is this supposed to exist? 

I know there's a system message for a failed SmartScan that should be added to the transaction thread that would notify the submitter when it fails. It looks like this: 

<img width="822" alt="image" src="https://github.com/user-attachments/assets/fdca9e6c-16fa-494e-942f-977ccad47885">

But interestingly, I just tested that, and it seems to not be added anymore? 🤔 

<img width="833" alt="image" src="https://github.com/user-attachments/assets/3100985b-875c-4bef-850a-dcab04c6002c">

>Secondly, is the desired behaviour to only send push notifications for the experience the user is set to? For example, a NewDot user would not receive a notification for an OldDot violation, and a NewDot user would not (currently) receive any violation notification.

On this, yes.. if you're set to the NewDot experience you shouldn't receive push notifications for the OldDot experience and the same the other way around. Otherwise, you would get pulled into the wrong experience when tapping the notification.

**Comment by trjExpensify** [comment]
Code for SmartScan failure is here, I believe: https://github.com/Expensify/Auth/blob/44606fb94035b15742c95973346b3d468269e4bc/auth/command/EditMoneyRequest.cpp#L431-L445

**Comment by Julesssss** [comment]
> What's a "NewDot violation", and what's an "OldDot violation", just so I'm clear?

I wasn't sure where they originated from (are the OldDot notifications coming from a NewDot violation request?) -- but your answer has cleared this up. 

> On this, yes.. if you're set to the NewDot experience you shouldn't receive push notifications for the OldDot experience and the same the other way around.

Okay thanks. Our logic assumes otherwise as I thought we wanted to avoid dropping OldDot notifications when users first switch to NewDot. Currently we'll send OldDot styled notifications (receipt needs approval, violation) to NewDot users and we can simply switch this off. 

Do we have any issues or plans to re-introduce NewDot notifications though? As we ramp up NewDot redirection users are going to be blissfully unaware of violations, ect.

**Comment by trjExpensify** [comment]
Yeah, I can't see any. So I think what we need to do is: 

- Get that actionable system message when SmartScan fails sending again (which will then send a push notification to the submitter and open the transaction thread) 
- Send an actionable system message when there's a policy violation (which will then send a push notification to the submitter and open the transaction thread) 
- Keep sending the Expensify Card push notification(s) at the point of sale, but if you're set to the NewDot experience, open the camera/transaction view in NewDot. 

Do you guys agree with that?

**Comment by JmillsExpensify** [comment]
That sounds good to me as a starting point.

**Comment by trjExpensify** [comment]
Alright, so what's the next step here and who's taking it on?

**Comment by Julesssss** [comment]
First lets solve the crash:

> Currently we'll send OldDot styled notifications (receipt needs approval, violation) to NewDot users and we can simply switch this off.

I'll happily do this, but need to finish up attendee tracking first.

**Comment by trjExpensify** [comment]
Okay, does that block defaulting new sign-ups to the NewDot experience though? If so, I think it's higher priority than attendee tracking in terms of priority, as these final issues to get this done come into #convert now.

**Comment by Julesssss** [comment]
> Okay, does that block defaulting new sign-ups to the NewDot experience though?

Yeah, it probably should. But as both are critical WhatsNext tasks associated with deadlines, I think I should finish what I started before picking up this task -- attendee tracking has already missed one deadline, and this would risk missing a second.

**Comment by Julesssss** [comment]
After catching up with the new prioritisations, I can pick this up once I have wound down attendee tracking and left it in a good place.

**Comment by trjExpensify** [comment]
Great stuff!

**Comment by trjExpensify** [comment]
Jules, are you good with the plan of action here? Which I think is:

1. Solve the crash. Stop sending the OldDot styled notifications when there's a violation on the receipt uploaded when the user is defaulted to the NewDot experience.
2. Add the NewDot system messages to the transaction thread, so they notify the submitter incl. bringing back the SmartScan failure one.

**SmartScan failure:**
>**Expensify** Today at 10:49AM
>receipt scanning failed, please enter the details manually.

**Violation on receipt upload:**

>**Expensify** Today at 09:02AM
>that %merchant% receipt you just added needs your attention, please take a look. 👀 

3. Check Expensify Card "point of sale" notifications are still delivered and don't crash on open, make them open the camera/transaction view in the NewDot experience when applicable.

**Comment by Julesssss** [comment]
Yeah, this all seems fine. I was planning to solve the crash first (1), and plan the remaining tasks (2,3) for later. But I could continue to working on these if we're okay delaying DevX/making HybridApp external. A couple of clarifications:

> Add the NewDot system messages to the transaction thread

I might have to configure the notification title/grouping. This would be minor text formatting though.

> Check Expensify Card "point of sale" notifications are still delivered

To be 100% sure, we still will send the OldDot POS notification to users on NewDot but won't yet build out the NewDot POS notifications.

**Comment by trjExpensify** [comment]
>I might have to configure the notification title/grouping. This would be minor text formatting though.

We had a system message for SmartScan failures already that we should bring back (or figure out why it's not there anymore), so I don't think this would be too hard to follow for this NewDot paradigm. 

>But I could continue to working on these if we're okay delaying DevX/making HybridApp external.

I think these items are probably more important as they impact customers using the NewDot experience on HybridApp. Ecard transactions POS notifications have a fraud component to them as well really, so I think this is the right priority personally. 👍 

>To be 100% sure, we still will send the OldDot POS notification to users on NewDot but won't yet build out the NewDot POS notifications.

Yeah, I view the Expensify Card POS notifications as being a feature of the Expensify Card not OldDot specifically. If you receive one, it should open in the right experience depending on which one that is.

**Comment by Julesssss** [comment]
Focusing on a more critical HybridApp issue first.

**Comment by Julesssss** [comment]
Picking this up again as the higher-priority HybridApp task was resolved.

**Comment by Julesssss** [comment]
I'm struggling a bit to reproduce. It seems that not all violations trigger the notification or cause the crash. 

**WIP Steps**
- Create NewDot account
- Create workspace
- Upgrade and enable 'Rules', set low receipt required limit
- `NVP.set('tryNewDot', {classicRedirect: {dismissed: "true"}});`
- On HybridApp ensure app is set to OldDot
- On Web upload bad receipt
- OldDot notification received 
- Tap notification
- Expense shown in OldDot and doesn't crash 😕

**Comment by Julesssss** [comment]
I'm App deployer this week so progress is likely to be slow. In the meantime, I am asking QA to reproduce and provide detailed reproduction steps.

**Comment by kavimuru** [comment]
No crash, but the notification does not lead to ND, OD with an error instead

https://github.com/user-attachments/assets/3cd8c441-b686-48ed-a30d-abb613456f0f

**Comment by trjExpensify** [comment]
>Create NewDot account
>Create workspace
>Upgrade and enable 'Rules', set low receipt required limit
>NVP.set('tryNewDot', {classicRedirect: {dismissed: "true"}});
>On HybridApp ensure app is set to OldDot
>On Web upload bad receipt
>OldDot notification received
>Tap notification
>Expense shown in OldDot and doesn't crash 😕

Out of curiosity Jules, why did you set classicRedirect to `true` and set to OldDot here? Isn't the bug that you're set to NewDot and still receiving the OD notification which crashes when clicked. 🤔 

>Clicking expense violation notifications cause the iOS HybridApp to crash. NewDot users should not be receiving OldDot notifications.

**Comment by Julesssss** [comment]
I'll need to try again, but I believe I tried both cases -- Initially I wasn't receiving the notification when set to NewDot, so I switched to these steps.

**Comment by Julesssss** [comment]
Thanks @kavimuru. So as of right now the crash isn't reproducable. Please anyone correct me if they disagree.

I'm too busy with daily App Deploy to pick this up yet but I my next step will be to make sure NewDot users aren't sent OldDot violation notifications.

> 1. ~~Solve the crash.~~ Stop sending the OldDot styled notifications when there's a violation on the receipt uploaded when the user is defaulted to the NewDot experience.

@trjExpensify if the crash isn't occurring then there's nothing blocking our *final HybridApp production deploy.

**Comment by trjExpensify** [comment]
>@trjExpensify if the crash isn't occurring then there's nothing blocking our *final HybridApp production deploy.

That's awesome news, let's confirm and ship!

**Comment by Julesssss** [comment]
I confirmed that this is no longer crashing for Andrew or myself. [Here's](https://github.com/Expensify/Expensify/issues/442751) the final HybridApp release for reference.

**Comment by Julesssss** [comment]
Finishing up other tasks. The priority might come below other tasks now that the crash isn't occurring.

**Comment by trjExpensify** [comment]
This can be closed now. The crash was fixed. Jason is progressing Ecard notifications elsewhere.
